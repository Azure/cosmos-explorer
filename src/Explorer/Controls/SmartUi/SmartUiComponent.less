@import "../../../../less/Common/Constants.less";

.widgetRendererContainer {
  text-align: left;

  .inputLabelContainer {
    margin-bottom: 4px;

    .inputLabel {
      color: #393939;
      font-weight: 600;
    }
  }
}