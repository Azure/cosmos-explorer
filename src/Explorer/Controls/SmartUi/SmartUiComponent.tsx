import * as React from "react";
import { Position } from "office-ui-fabric-react/lib/utilities/positioning";
import { Slider } from "office-ui-fabric-react/lib/Slider";
import { SpinButton } from "office-ui-fabric-react/lib/SpinButton";
import { Dropdown, IDropdownOption } from "office-ui-fabric-react/lib/Dropdown";
import { TextField } from "office-ui-fabric-react/lib/TextField";
import { Text } from "office-ui-fabric-react/lib/Text";
import { RadioSwitchComponent } from "../RadioSwitchComponent/RadioSwitchComponent";
import { Stack, IStackTokens } from "office-ui-fabric-react/lib/Stack";
import { Link, MessageBar, MessageBarType } from "office-ui-fabric-react";
import * as InputUtils from "./InputUtils";
import "./SmartUiComponent.less";

/**
 * Generic UX renderer
 * It takes:
 * - a JSON object as data
 * - a Map of callbacks
 * - a descriptor of the UX.
 */

export type InputTypeValue = "number" | "string" | "boolean" | "object";

export enum UiType {
  Spinner = "Spinner",
  Slider = "Slider"
}

export type ChoiceItem = { label: string; key: string };

export type InputType = number | string | boolean | ChoiceItem;

export interface Info {
  message: string;
  link?: {
    href: string;
    text: string;
  };
}

interface BaseInput {
  label: string;
  dataFieldName: string;
  type: InputTypeValue;
  placeholder?: string;
  errorMessage?: string;
}

/**
 * For now, this only supports integers
 */
interface NumberInput extends BaseInput {
  min: number;
  max: number;
  step: number;
  defaultValue?: number;
  uiType: UiType;
}

interface BooleanInput extends BaseInput {
  trueLabel: string;
  falseLabel: string;
  defaultValue?: boolean;
}

interface StringInput extends BaseInput {
  defaultValue?: string;
}

interface ChoiceInput extends BaseInput {
  choices: ChoiceItem[];
  defaultKey?: string;
}

type AnyInput = NumberInput | BooleanInput | StringInput | ChoiceInput;

interface Node {
  id: string;
  info?: Info;
  input?: AnyInput;
  children?: Node[];
}

export interface SmartUiDescriptor {
  root: Node;
}

/************************** Component implementation starts here ************************************* */

export interface SmartUiComponentProps {
  descriptor: SmartUiDescriptor;
  currentValues: Map<string, InputType>;
  onInputChange: (input: AnyInput, newValue: InputType) => void;
}

interface SmartUiComponentState {
  errors: Map<string, string>;
}

export class SmartUiComponent extends React.Component<SmartUiComponentProps, SmartUiComponentState> {
  private static readonly labelStyle = {
    color: "#393939",
    fontFamily: "wf_segoe-ui_normal, 'Segoe UI', 'Segoe WP', Tahoma, Arial, sans-serif",
    fontSize: 12
  };

  constructor(props: SmartUiComponentProps) {
    super(props);
    this.state = {
      errors: new Map()
    };
  }

  private renderInfo(info: Info): JSX.Element {
    return (
      <MessageBar>
        {info.message}
        {info.link && (
          <Link href={info.link.href} target="_blank">
            {info.link.text}
          </Link>
        )}
      </MessageBar>
    );
  }

  private renderTextInput(input: StringInput): JSX.Element {
    const value = this.props.currentValues.get(input.dataFieldName) as string;
    return (
      <div className="stringInputContainer">
        <TextField
          id={`${input.dataFieldName}-textBox-input`}
          label={input.label}
          type="text"
          value={value}
          placeholder={input.placeholder}
          onChange={(_, newValue) => this.props.onInputChange(input, newValue)}
          styles={{
            subComponentStyles: {
              label: {
                root: {
                  ...SmartUiComponent.labelStyle,
                  fontWeight: 600
                }
              }
            }
          }}
        />
      </div>
    );
  }

  private clearError(dataFieldName: string): void {
    const { errors } = this.state;
    errors.delete(dataFieldName);
    this.setState({ errors });
  }

  private onValidate = (input: AnyInput, value: string, min: number, max: number): string => {
    const newValue = InputUtils.onValidateValueChange(value, min, max);
    const dataFieldName = input.dataFieldName;
    if (newValue) {
      this.props.onInputChange(input, newValue);
      this.clearError(dataFieldName);
      return newValue.toString();
    } else {
      const { errors } = this.state;
      errors.set(dataFieldName, `Invalid value ${value}: must be between ${min} and ${max}`);
      this.setState({ errors });
    }
    return undefined;
  };

  private onIncrement = (input: AnyInput, value: string, step: number, max: number): string => {
    const newValue = InputUtils.onIncrementValue(value, step, max);
    const dataFieldName = input.dataFieldName;
    if (newValue) {
      this.props.onInputChange(input, newValue);
      this.clearError(dataFieldName);
      return newValue.toString();
    }
    return undefined;
  };

  private onDecrement = (input: AnyInput, value: string, step: number, min: number): string => {
    const newValue = InputUtils.onDecrementValue(value, step, min);
    const dataFieldName = input.dataFieldName;
    if (newValue) {
      this.props.onInputChange(input, newValue);
      this.clearError(dataFieldName);
      return newValue.toString();
    }
    return undefined;
  };

  private renderNumberInput(input: NumberInput): JSX.Element {
    const { label, min, max, dataFieldName, step } = input;
    const props = {
      label: label,
      min: min,
      max: max,
      ariaLabel: label,
      step: step
    };

    const value = this.props.currentValues.get(dataFieldName) as number;
    if (input.uiType === UiType.Spinner) {
      return (
        <>
          <SpinButton
            {...props}
            id={`${input.dataFieldName}-spinner-input`}
            value={value?.toString()}
            onValidate={newValue => this.onValidate(input, newValue, props.min, props.max)}
            onIncrement={newValue => this.onIncrement(input, newValue, props.step, props.max)}
            onDecrement={newValue => this.onDecrement(input, newValue, props.step, props.min)}
            labelPosition={Position.top}
            styles={{
              label: {
                ...SmartUiComponent.labelStyle,
                fontWeight: 600
              }
            }}
          />
          {this.state.errors.has(dataFieldName) && (
            <MessageBar messageBarType={MessageBarType.error}>Error: {this.state.errors.get(dataFieldName)}</MessageBar>
          )}
        </>
      );
    } else if (input.uiType === UiType.Slider) {
      return (
        <div id={`${input.dataFieldName}-slider-input`}>
          <Slider
            {...props}
            value={value}
            onChange={newValue => this.props.onInputChange(input, newValue)}
            styles={{
              titleLabel: {
                ...SmartUiComponent.labelStyle,
                fontWeight: 600
              },
              valueLabel: SmartUiComponent.labelStyle
            }}
          />
        </div>
      );
    } else {
      return <>Unsupported number UI type {input.uiType}</>;
    }
  }

  private renderBooleanInput(input: BooleanInput): JSX.Element {
    const value = this.props.currentValues.get(input.dataFieldName) as boolean;
    const selectedKey = value || input.defaultValue ? "true" : "false";
    return (
      <div id={`${input.dataFieldName}-radioSwitch-input`}>
        <div className="inputLabelContainer">
          <Text variant="small" nowrap className="inputLabel">
            {input.label}
          </Text>
        </div>
        <RadioSwitchComponent
          choices={[
            {
              label: input.falseLabel,
              key: "false",
              onSelect: () => this.props.onInputChange(input, false)
            },
            {
              label: input.trueLabel,
              key: "true",
              onSelect: () => this.props.onInputChange(input, true)
            }
          ]}
          selectedKey={selectedKey}
        />
      </div>
    );
  }

  private renderChoiceInput(input: ChoiceInput): JSX.Element {
    const { label, defaultKey: defaultKey, dataFieldName, choices, placeholder } = input;
    const value = this.props.currentValues.get(dataFieldName) as string;
    return (
      <Dropdown
        id={`${input.dataFieldName}-dropown-input`}
        label={label}
        selectedKey={value ? value : defaultKey}
        onChange={(_, item: IDropdownOption) => this.props.onInputChange(input, item.key.toString())}
        placeholder={placeholder}
        options={choices.map(c => ({
          key: c.key,
          text: c.label
        }))}
        styles={{
          label: {
            ...SmartUiComponent.labelStyle,
            fontWeight: 600
          },
          dropdown: SmartUiComponent.labelStyle
        }}
      />
    );
  }

  private renderError(input: AnyInput): JSX.Element {
    return <MessageBar messageBarType={MessageBarType.error}>Error: {input.errorMessage}</MessageBar>;
  }

  private renderInput(input: AnyInput): JSX.Element {
    if (input.errorMessage) {
      return this.renderError(input);
    }
    switch (input.type) {
      case "string":
        return this.renderTextInput(input as StringInput);
      case "number":
        return this.renderNumberInput(input as NumberInput);
      case "boolean":
        return this.renderBooleanInput(input as BooleanInput);
      case "object":
        return this.renderChoiceInput(input as ChoiceInput);
      default:
        throw new Error(`Unknown input type: ${input.type}`);
    }
  }

  private renderNode(node: Node): JSX.Element {
    const containerStackTokens: IStackTokens = { childrenGap: 15 };

    return (
      <Stack tokens={containerStackTokens} className="widgetRendererContainer">
        <Stack.Item>
          {node.info && this.renderInfo(node.info as Info)}
          {node.input && this.renderInput(node.input)}
        </Stack.Item>
        {node.children && node.children.map(child => <div key={child.id}>{this.renderNode(child)}</div>)}
      </Stack>
    );
  }

  render(): JSX.Element {
    const containerStackTokens: IStackTokens = { childrenGap: 20 };
    return (
      <Stack tokens={containerStackTokens} styles={{ root: { width: 400, padding: 10 } }}>
        {this.renderNode(this.props.descriptor.root)}
      </Stack>
    );
  }
}
