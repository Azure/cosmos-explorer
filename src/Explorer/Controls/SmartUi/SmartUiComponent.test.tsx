import React from "react";
import { shallow } from "enzyme";
import { SmartUiComponent, SmartUiDescriptor, NumberUiType, SmartUiInput } from "./SmartUiComponent";

describe("SmartUiComponent", () => {
  const exampleData: SmartUiDescriptor = {
    root: {
      id: "root",
      info: {
        message: "Start at $24/mo per database",
        link: {
          href: "https://aka.ms/azure-cosmos-db-pricing",
          text: "More Details",
        },
      },
      children: [
        {
          id: "description",
          input: {
            label: undefined,
            dataFieldName: "description",
            type: "string",
            description: {
              text: "this is an example description text.",
              link: {
                href: "https://docs.microsoft.com/en-us/azure/cosmos-db/introduction",
                text: "Click here for more information.",
              },
            },
          },
        },
        {
          id: "throughput",
          input: {
            label: "Throughput (input)",
            dataFieldName: "throughput",
            type: "number",
            min: 400,
            max: 500,
            step: 10,
            defaultValue: 400,
            uiType: NumberUiType.Spinner,
          },
        },
        {
          id: "throughput2",
          input: {
            label: "Throughput (Slider)",
            dataFieldName: "throughput2",
            type: "number",
            min: 400,
            max: 500,
            step: 10,
            defaultValue: 400,
            uiType: NumberUiType.Slider,
          },
        },
        {
          id: "throughput3",
          input: {
            label: "Throughput (invalid)",
            dataFieldName: "throughput3",
            type: "boolean",
            min: 400,
            max: 500,
            step: 10,
            defaultValue: 400,
            uiType: NumberUiType.Spinner,
            errorMessage: "label, truelabel and falselabel are required for boolean input 'throughput3'",
          },
        },
        {
          id: "containerId",
          input: {
            label: "Container id",
            dataFieldName: "containerId",
            type: "string",
          },
        },
        {
          id: "analyticalStore",
          input: {
            label: "Analytical Store",
            trueLabel: "Enabled",
            falseLabel: "Disabled",
            defaultValue: true,
            dataFieldName: "analyticalStore",
            type: "boolean",
          },
        },
        {
          id: "database",
          input: {
            label: "Database",
            dataFieldName: "database",
            type: "object",
            choices: [
              { label: "Database 1", key: "db1" },
              { label: "Database 2", key: "db2" },
              { label: "Database 3", key: "db3" },
            ],
            defaultKey: "db2",
          },
        },
      ],
    },
  };

  it("should render and honor input's hidden, disabled state", async () => {
    const currentValues = new Map<string, SmartUiInput>();
    const wrapper = shallow(
      <SmartUiComponent
        disabled={false}
        descriptor={exampleData}
        currentValues={currentValues}
        onInputChange={undefined}
        onError={() => {
          return;
        }}
      />
    );
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.exists("#containerId-textField-input")).toBeTruthy();

    currentValues.set("containerId", { value: "container1", hidden: true });
    wrapper.setProps({ currentValues });
    wrapper.update();
    expect(wrapper.exists("#containerId-textField-input")).toBeFalsy();

    currentValues.set("containerId", { value: "container1", hidden: false, disabled: true });
    wrapper.setProps({ currentValues });
    wrapper.update();
    const containerIdTextField = wrapper.find("#containerId-textField-input");
    expect(containerIdTextField.props().disabled).toBeTruthy();
  });

  it("disable all inputs", async () => {
    const wrapper = shallow(
      <SmartUiComponent
        disabled={true}
        descriptor={exampleData}
        currentValues={new Map()}
        onInputChange={undefined}
        onError={() => {
          return;
        }}
      />
    );
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(wrapper).toMatchSnapshot();
    const throughputSpinner = wrapper.find("#throughput-spinner-input");
    expect(throughputSpinner.props().disabled).toBeTruthy();
    const throughput2Slider = wrapper.find("#throughput2-slider-input").childAt(0);
    expect(throughput2Slider.props().disabled).toBeTruthy();
    const containerIdTextField = wrapper.find("#containerId-textField-input");
    expect(containerIdTextField.props().disabled).toBeTruthy();
    const analyticalStoreToggle = wrapper.find("#analyticalStore-toggle-input");
    expect(analyticalStoreToggle.props().disabled).toBeTruthy();
    const databaseDropdown = wrapper.find("#database-dropdown-input");
    expect(databaseDropdown.props().disabled).toBeTruthy();
  });
});
