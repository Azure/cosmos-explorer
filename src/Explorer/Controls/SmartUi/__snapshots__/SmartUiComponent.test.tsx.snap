// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartUiComponent disable all inputs 1`] = `
<Stack
  className="widgetRendererContainer"
  tokens={
    Object {
      "childrenGap": 10,
    }
  }
>
  <StackItem />
  <div
    key="description"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <Text
              aria-labelledby="description-label"
              id="description-text-display"
            >
              this is an example description text.
               
              <StyledLinkBase
                href="https://docs.microsoft.com/en-us/azure/cosmos-db/introduction"
                target="_blank"
              >
                Click here for more information.
              </StyledLinkBase>
            </Text>
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="throughput"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="throughput-label"
            >
              <ToolTipLabelComponent
                label="Throughput (input)"
              />
            </StyledLabelBase>
            <Stack
              styles={
                Object {
                  "root": Object {
                    "width": 400,
                  },
                }
              }
              tokens={
                Object {
                  "childrenGap": 2,
                }
              }
            >
              <CustomizedSpinButton
                aria-labelledby="throughput-label"
                ariaLabel="Throughput (input)"
                decrementButtonIcon={
                  Object {
                    "iconName": "ChevronDownSmall",
                  }
                }
                disabled={true}
                id="throughput-spinner-input"
                incrementButtonIcon={
                  Object {
                    "iconName": "ChevronUpSmall",
                  }
                }
                label=""
                labelPosition={0}
                max={500}
                min={400}
                onDecrement={[Function]}
                onIncrement={[Function]}
                onValidate={[Function]}
                step={10}
              />
            </Stack>
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="throughput2"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="throughput2-label"
            >
              <ToolTipLabelComponent
                label="Throughput (Slider)"
              />
            </StyledLabelBase>
            <div
              id="throughput2-slider-input"
            >
              <StyledSliderBase
                ariaLabel="Throughput (Slider)"
                disabled={true}
                max={500}
                min={400}
                onChange={[Function]}
                step={10}
                styles={
                  Object {
                    "root": Object {
                      "width": 400,
                    },
                    "valueLabel": Object {
                      "color": "#393939",
                      "fontFamily": "wf_segoe-ui_normal, 'Segoe UI', 'Segoe WP', Tahoma, Arial, sans-serif",
                      "fontSize": 12,
                    },
                  }
                }
              />
            </div>
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="throughput3"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <StyledMessageBarBase
          messageBarType={1}
        >
          Error: 
          label, truelabel and falselabel are required for boolean input 'throughput3'
        </StyledMessageBarBase>
      </StackItem>
    </Stack>
  </div>
  <div
    key="containerId"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="containerId-label"
            >
              <ToolTipLabelComponent
                label="Container id"
              />
            </StyledLabelBase>
            <StyledTextFieldBase
              aria-labelledby="containerId-label"
              disabled={true}
              id="containerId-textField-input"
              onChange={[Function]}
              styles={
                Object {
                  "root": Object {
                    "width": 400,
                  },
                }
              }
              type="text"
              value=""
            />
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="analyticalStore"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="analyticalStore-label"
            >
              <ToolTipLabelComponent
                label="Analytical Store"
              />
            </StyledLabelBase>
            <StyledToggleBase
              aria-labelledby="analyticalStore-label"
              checked={false}
              disabled={true}
              id="analyticalStore-toggle-input"
              offText="Disabled"
              onChange={[Function]}
              onText="Enabled"
              styles={
                Object {
                  "root": Object {
                    "width": 400,
                  },
                }
              }
            />
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="database"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="database-label"
            >
              <ToolTipLabelComponent
                label="Database"
              />
            </StyledLabelBase>
            <StyledWithResponsiveMode
              aria-labelledby="database-label"
              disabled={true}
              id="database-dropdown-input"
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "key": "db1",
                    "text": "Database 1",
                  },
                  Object {
                    "key": "db2",
                    "text": "Database 2",
                  },
                  Object {
                    "key": "db3",
                    "text": "Database 3",
                  },
                ]
              }
              selectedKey="db2"
              styles={
                Object {
                  "dropdown": Object {
                    "color": "#393939",
                    "fontFamily": "wf_segoe-ui_normal, 'Segoe UI', 'Segoe WP', Tahoma, Arial, sans-serif",
                    "fontSize": 12,
                  },
                  "root": Object {
                    "width": 400,
                  },
                }
              }
            />
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
</Stack>
`;

exports[`SmartUiComponent should render and honor input's hidden, disabled state 1`] = `
<Stack
  className="widgetRendererContainer"
  tokens={
    Object {
      "childrenGap": 10,
    }
  }
>
  <StackItem />
  <div
    key="description"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <Text
              aria-labelledby="description-label"
              id="description-text-display"
            >
              this is an example description text.
               
              <StyledLinkBase
                href="https://docs.microsoft.com/en-us/azure/cosmos-db/introduction"
                target="_blank"
              >
                Click here for more information.
              </StyledLinkBase>
            </Text>
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="throughput"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="throughput-label"
            >
              <ToolTipLabelComponent
                label="Throughput (input)"
              />
            </StyledLabelBase>
            <Stack
              styles={
                Object {
                  "root": Object {
                    "width": 400,
                  },
                }
              }
              tokens={
                Object {
                  "childrenGap": 2,
                }
              }
            >
              <CustomizedSpinButton
                aria-labelledby="throughput-label"
                ariaLabel="Throughput (input)"
                decrementButtonIcon={
                  Object {
                    "iconName": "ChevronDownSmall",
                  }
                }
                disabled={false}
                id="throughput-spinner-input"
                incrementButtonIcon={
                  Object {
                    "iconName": "ChevronUpSmall",
                  }
                }
                label=""
                labelPosition={0}
                max={500}
                min={400}
                onDecrement={[Function]}
                onIncrement={[Function]}
                onValidate={[Function]}
                step={10}
              />
            </Stack>
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="throughput2"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="throughput2-label"
            >
              <ToolTipLabelComponent
                label="Throughput (Slider)"
              />
            </StyledLabelBase>
            <div
              id="throughput2-slider-input"
            >
              <StyledSliderBase
                ariaLabel="Throughput (Slider)"
                max={500}
                min={400}
                onChange={[Function]}
                step={10}
                styles={
                  Object {
                    "root": Object {
                      "width": 400,
                    },
                    "valueLabel": Object {
                      "color": "#393939",
                      "fontFamily": "wf_segoe-ui_normal, 'Segoe UI', 'Segoe WP', Tahoma, Arial, sans-serif",
                      "fontSize": 12,
                    },
                  }
                }
              />
            </div>
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="throughput3"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <StyledMessageBarBase
          messageBarType={1}
        >
          Error: 
          label, truelabel and falselabel are required for boolean input 'throughput3'
        </StyledMessageBarBase>
      </StackItem>
    </Stack>
  </div>
  <div
    key="containerId"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="containerId-label"
            >
              <ToolTipLabelComponent
                label="Container id"
              />
            </StyledLabelBase>
            <StyledTextFieldBase
              aria-labelledby="containerId-label"
              id="containerId-textField-input"
              onChange={[Function]}
              styles={
                Object {
                  "root": Object {
                    "width": 400,
                  },
                }
              }
              type="text"
              value=""
            />
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="analyticalStore"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="analyticalStore-label"
            >
              <ToolTipLabelComponent
                label="Analytical Store"
              />
            </StyledLabelBase>
            <StyledToggleBase
              aria-labelledby="analyticalStore-label"
              checked={false}
              id="analyticalStore-toggle-input"
              offText="Disabled"
              onChange={[Function]}
              onText="Enabled"
              styles={
                Object {
                  "root": Object {
                    "width": 400,
                  },
                }
              }
            />
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
  <div
    key="database"
  >
    <Stack
      className="widgetRendererContainer"
      tokens={
        Object {
          "childrenGap": 10,
        }
      }
    >
      <StackItem>
        <Stack>
          <Stack>
            <StyledLabelBase
              id="database-label"
            >
              <ToolTipLabelComponent
                label="Database"
              />
            </StyledLabelBase>
            <StyledWithResponsiveMode
              aria-labelledby="database-label"
              id="database-dropdown-input"
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "key": "db1",
                    "text": "Database 1",
                  },
                  Object {
                    "key": "db2",
                    "text": "Database 2",
                  },
                  Object {
                    "key": "db3",
                    "text": "Database 3",
                  },
                ]
              }
              selectedKey="db2"
              styles={
                Object {
                  "dropdown": Object {
                    "color": "#393939",
                    "fontFamily": "wf_segoe-ui_normal, 'Segoe UI', 'Segoe WP', Tahoma, Arial, sans-serif",
                    "fontSize": 12,
                  },
                  "root": Object {
                    "width": 400,
                  },
                }
              }
            />
          </Stack>
        </Stack>
      </StackItem>
    </Stack>
  </div>
</Stack>
`;
